rclpy/src/rclpy/node.cpp

```diff
@@ -368,4 +372,139 @@ get_node_parameters(py::object pyparameter_cls, py::capsule pynode)
+
+  rcl_ret_t ret = rcl_node_fini(node);
+  if (RCL_RET_OK != ret) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini node: %s",
+      rcl_get_error_string().str);
+  }
+  PyMem_Free(node);
+}
+
+py::capsule
+create_node(
+  const char * node_name,
+  const char * namespace_,
+  py::capsule pycontext,
+  py::object pycli_args,
+  bool use_global_arguments,
+  bool enable_rosout)
+{
+  rcl_ret_t ret;
+
+  auto context = static_cast<rcl_context_t *>(
+    rclpy_handle_get_pointer_from_capsule(pycontext.ptr(), "rcl_context_t"));
+  if (!context) {
+    throw py::error_already_set();
+  }
+
+  rcl_arguments_t arguments = rcl_get_zero_initialized_arguments();
+
+  // turn the arguments into an array of C-style strings
+  std::vector<const char *> arg_values;
+  const char ** const_arg_values = NULL;
+  py::list pyargs;
+  if (!pycli_args.is_none()) {
+    pyargs = pycli_args;
+    arg_values.resize(pyargs.size());
+    for (size_t i = 0; i < pyargs.size(); ++i) {
+      // CPython owns const char * memory - no need to free it
+      arg_values[i] = PyUnicode_AsUTF8(pyargs[i].ptr());
+      if (!arg_values[i]) {
+        throw py::error_already_set();
+      }
+    }
+    const_arg_values = &(arg_values[0]);
+  }
+
+  // call rcl_parse_arguments() so rcl_arguments_t structure is always valid.
+  // Otherwise the remapping functions will error if the user passes no arguments to a node and sets
+  // use_global_arguments to False.
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  ret = rcl_parse_arguments(arg_values.size(), const_arg_values, allocator, &arguments);
+
+  if (RCL_RET_INVALID_ROS_ARGS == ret) {
+    throw RCLInvalidROSArgsError("Failed to parse ROS arguments");
+  }
+  if (RCL_RET_OK != ret) {
+    throw RCLError("failed to parse arguments");
+  }
+
+  RCPPUTILS_SCOPE_EXIT(
+    {
+      if (RCL_RET_OK != rcl_arguments_fini(&arguments)) {
+        int stack_level = 1;
+        PyErr_WarnFormat(
+          PyExc_RuntimeWarning, stack_level, "Failed to fini arguments: %s",
+          rcl_get_error_string().str);
+        rcl_reset_error();
+      }
+    });
+
+  throw_if_unparsed_ros_args(pyargs, arguments);
+
+  auto deleter = [](rcl_node_t * ptr) {_rclpy_destroy_node(ptr);};
+  auto node = std::unique_ptr<rcl_node_t, decltype(deleter)>(
+    static_cast<rcl_node_t *>(PyMem_Malloc(sizeof(rcl_node_t))),
+    deleter);
+  if (!node) {
+    throw std::bad_alloc();
+  }
+  *node = rcl_get_zero_initialized_node();
+  rcl_node_options_t options = rcl_node_get_default_options();
+  options.use_global_arguments = use_global_arguments;
+  options.arguments = arguments;
+  options.enable_rosout = enable_rosout;
+
+  {
+    rclpy::LoggingGuard scoped_logging_guard;
+    ret = rcl_node_init(node.get(), node_name, namespace_, context, &options);
+  }
+
+  if (RCL_RET_BAD_ALLOC == ret) {
+    throw std::bad_alloc();
```

Comment:
@sloretz consider doing `rcl_reset_error()` before throwing.

Reply:
Reset the rcl error in 336ca4f5803cb5fe4e1c6a88e9ea12556d115a57